runtime/include/tt/runtime/detail/ttnn/types/global_tensor_cache.h

```diff
@@ -0,0 +1,104 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TT_RUNTIME_DETAIL_TTNN_TYPES_GLOBAL_TENSOR_CACHE_H
+#define TT_RUNTIME_DETAIL_TTNN_TYPES_GLOBAL_TENSOR_CACHE_H
+
+#include "tt/runtime/types.h"
+
+#include <cstdint>
+#include <functional>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace tt::runtime {
+struct CacheKey {
+  int deviceId;
+  std::string funcHash;
+  std::vector<uint64_t> inputVersions;
+
+  CacheKey() = default;
+  CacheKey(int deviceId, const std::string funcHash,
+           const std::vector<uint64_t> inputVersions)
+      : deviceId(deviceId), funcHash(std::move(funcHash)),
+        inputVersions(std::move(inputVersions)) {}
+
+  bool operator==(const CacheKey &other) const {
+    return deviceId == other.deviceId && funcHash == other.funcHash &&
+           inputVersions == other.inputVersions;
+  }
+};
+
+std::string to_string(const CacheKey &key);
+
+struct CacheKeyHasher {
+  std::size_t operator()(const CacheKey &k) const {
+    // TODO: Optimize hash function if needed
```

Comment:
:warning: **google\-readability\-todo** :warning:
missing username/bug in TODO
```suggestion
    // TODO(nsmith): Optimize hash function if needed
```